File: ruby/day6/day6.rb
require 'set'

p1 = 0
p2 = 0
for group in File.read('input').split("\n\n")
  s = Set.new

  person_count = group.split("\n").size
  yeps = {}
  for answer in group.split("\n")
    answer.chars.each { |c| s.add(c) }
    answer.chars.each { |c| yeps.key?(c) ? yeps[c] += 1 : yeps[c] = 1 }
  end

  p1 += s.size
  p2 += yeps.values.select { |n| n == person_count }.size
end

p p1
p p2

File: haskell/day6/day6.hs
import Data.List.Split
import Data.List
import qualified Data.Map as M

type Group = [String]
type AnswerMap = M.Map Char Int

parseInput :: String -> [Group]
parseInput raw = map lines $ groups
  where groups = splitOn "\n\n" raw

part1 :: [Group] -> Int
part1 groups = sum $ map yesPerGroup groups
    where yesPerGroup = length . nub . foldl1 (++)

-- part2 :: [Group] -> Int
part2 groups = (head groups)
    where addAnswers :: AnswerMap -> String -> AnswerMap
          addAnswers am answers = foldl (\c m -> M.insertWith' (+) c 1 m) mm

answersMap :: M.Map Char Int
answersMap = M.fromList [(c, 0) | c <- ['a'..'z']]

main = do
    raw_input <- readFile "input"
    let groups = parseInput raw_input

    -- print $ part1 groups
    print $ part2 groups
